source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -339,18 +339,144 @@ state and balance fields:
+
+.. code-block:: rust
+
+    #[concordium_test]
+    fn test_reentrant_withdraw() {
+        ...
+
+        // Assume that `CONTRACT_ADDRESS` has 1 micro CCD
+        // Set the contract balance to 2 micro CCD
+        host.set_self_balance(Amount::from_micro_ccd(2));
+
+        // Set up a mock entrypoint that calls back to our contract.
+        // We emulate the `reentrant_withdraw` logic here.
+        host.setup_mock_entrypoint(
+            CONTRACT_ADDRESS,
+            OwnedEntrypointName::new_unchecked("reentrant_withdraw".to_string()),
+            MockFn::new_v1(|_parameter, _amount, balance, state: &mut State<_>| {
+                // We cannot call `invoke_contract` inside this mock, but we have
+                // access to the `balance`, which is the balance of the contract
+                // making this invocation. So we can simulate invoking withdraw by
+                // subtracting the sender's amount stored in the contract state
+                // from the balance.
+
+                let b = state.balances.get_mut(&Address::Contract(CONTRACT_ADDRESS));
+
+                let mut sender_balance = match b {
+                    Some(bal) if *bal > Amount::zero() => bal,
+                    _ => fail!("Insufficent funds"),
+                };
+
+                // Emulate withdraw by subtracting the sender's balance.
+                *balance -= *sender_balance;
+
+                // Reset the sender's balance to zero.
+                *sender_balance = Amount::zero();
+
+                let state_modified = true;
+                Ok((state_modified, ()))
+            }),
+        );
+        // Withdraw 1 micro CCD
+        reentrant_withdraw(&ctx, &mut host).expect_report("Withdraw call failed");
+
+        let resulting_balance = host.self_balance();
+        let expected_balance = 1;
+
+        claim_eq!(
+            resulting_balance,
+            expected_balance,
+            "Balance is not updated correctly: expected {:?}, found: {:?}",
+            expected_balance,
+            resulting_balance
+        );
+    }
 
-   You should watch out for *reentrancy problems*, which can occur when calls to
-   ``invoke_contract`` end up updating the state of your own contract.
+The test fails with the following message:
 
-   .. code-block:: rust
+.. code-block:: text
 
-      let state_copy = *host.state();
-      host.invoke_contract(...);
+    Incorrect balance: expected Amount { micro_ccd: 1 }, found: Amount { micro_ccd: 0 }
 
-      // *host.state() and state_copy might not be equal any more due to reentrancy.
-      do_something_with(state_copy);
+That means that the contract we called has stolen our funds through a reentrant call.
+A simple fix to this behavior is to place the highlighted line in ``reentrant_withdraw`` *before* making a call to an external contract.
+In this case we expect the ``reentrant_withdraw`` call to fail because the non-zero balance condition is not longer satisfied in the mock entrypoint.
```

Comment:
```suggestion
In this case the ``reentrant_withdraw`` call will fail because the non-zero balance condition is no longer satisfied in the mock entrypoint.
```